root/tree/dataframe/zpeak.cxx

```diff
@@ -0,0 +1,165 @@
+
+   // find 2 leptons with highest pt (electrons have precedence over muons)
+   auto df_bestleps =
+      df.Define("best_electrons", FindBestMaybe,
+                {"Electron_pt", "Electron_eta", "Electron_phi", "Electron_mass", "Electron_charge", "good_el_idx"})
+         .Define("best_muons", FindBestMaybe,
+                 {"Muon_pt", "Muon_eta", "Muon_phi", "Muon_mass", "Muon_charge", "good_mu_idx"})
+         .Define("best_leptons", FindBestLeptons, {"best_muons", "best_electrons"});
+
+   // fill histograms
+   auto df_histvars =
+      df_bestleps
+         .Define("dilep_mass",
+                 [](const Best2Leptons &leps) { return (leps[0].LorentzVec() + leps[1].LorentzVec()).M(); },
+                 {"best_leptons"})
+         .Define("pt1", [](const Best2Leptons &leps) { return leps[0].LorentzVec().Pt(); }, {"best_leptons"})
+         .Define("pt2", [](const Best2Leptons &leps) { return leps[1].LorentzVec().Pt(); }, {"best_leptons"});
+
+   auto id_product_is = [](int prod) {
+      return [prod](const Best2Leptons &leps) { return leps[0].Id() * leps[1].Id() == prod; };
+   };
+
+   auto df_ee = df_histvars.Filter(id_product_is(-11 * 11), {"best_leptons"});
+   auto mass_ee = df_ee.Histo1D<double>({"mass_ee", "mass;M_{ll} [GeV];Counts", 120, 0., 120.}, "dilep_mass");
+   auto lepPt_ee = df_ee.Histo2D<double, double>(
+      {"lepPt_ee", "LepPt;Leading p_{T}^{l} [GeV];Trailing p_{T}^{l} [GeV];Counts", 50, 0., 500., 50, 0., 500.}, "pt1",
+      "pt2");
+
+   auto df_em = df_histvars.Filter(id_product_is(-11 * 13), {"best_leptons"});
+   auto mass_em = df_em.Histo1D<double>({"mass_em", "mass;M_{ll} [GeV];Counts", 120, 0., 120.}, "dilep_mass");
+   auto lepPt_em = df_em.Histo2D<double, double>(
+      {"lepPt_em", "LepPt;Leading p_{T}^{l} [GeV];Trailing p_{T}^{l} [GeV];Counts", 50, 0., 500., 50, 0., 500.}, "pt1",
+      "pt2");
+
+   auto df_mm = df_histvars.Filter(id_product_is(-13 * 13), {"best_leptons"});
+   auto mass_mm = df_mm.Histo1D<double>({"mass_mm", "mass;M_{ll} [GeV];Counts", 120, 0., 120.}, "dilep_mass");
+   auto lepPt_mm = df_mm.Histo2D<double, double>(
+      {"lepPt_mm", "LepPt;Leading p_{T}^{l} [GeV];Trailing p_{T}^{l} [GeV];Counts", 50, 0., 500., 50, 0., 500.}, "pt1",
+      "pt2");
+
+   mass_ee.GetValue(); // trigger event loop
+
+   // left here for debugging purposes
+   /*
+   std::unique_ptr<TFile> fOut{TFile::Open("plots_RDF_nojit.root", "recreate")};
+   mass_ee->Write();
+   mass_em->Write();
+   mass_mm->Write();
+   lepPt_ee->Write();
+   lepPt_em->Write();
+   lepPt_mm->Write();
+   */
+}
+
+static void BM_RDataFrame_ZPeak(benchmark::State &state) {
+   for (auto _ : state)
+      run_zpeak();
+}
+BENCHMARK(BM_RDataFrame_ZPeak)->Repetitions(1);
+
+static void BM_RDataFrame_ZPeak_MT(benchmark::State &state) {
+   ROOT::EnableImplicitMT(state.range(0));
+   for (auto _ : state)
+      run_zpeak();
+   ROOT::DisableImplicitMT();
+}
+BENCHMARK(BM_RDataFrame_ZPeak_MT)->Repetitions(1)->Arg(8);
```

Comment:
I've used for other IMT benchmarks the auto detection of `ROOT::EnableImplicitMT()`. What should we do in general?

Reply:
I think that I agree that we should fix the number of cores used for IMT. We may move to another machines with more/less cores.

Reply:
Also we saturate IO with 56 cores completely. Probably we want to see actual scaling for the benchmarks!

Reply:
a machine with less cores shouldn't be a problem, right?

Reply:
In any case, you shouldn't compare between machines. If you were to compare the times from running in the KNL at 1.2 GHz and running on an AMD EPYC 7XXX with the same number of threads, times would be wildly different

Reply:
so, do you think 8 threads is ok or should we set it to 4? I don't have a strong opinion. I agree with Stefan that leaving it auto-detected might be source of confusion (and we don't want to run with 56 cores in nightly rootbench jobs anyway).

Reply:
You can check if you can just put a max boundary. You might get warnings, but 8/32/56 threads in a max 4 threads SMT processor (my laptop, f.e.) should not be a problem as TBB takes care of that and will run with 4.  Obviously, as long as you poll for max concurrency (ROOT::GetThreadPoolSize())  instead of hardcoding the number of threads if you want to return metrics that take the concurrency level into account 🙃

Reply:
uhm i'd just like to pick a number between 4 and 8, is that problematic?

Reply:
I doubt it